### PR TITLE
Add a hook for idle session killing

### PR DESCRIPTION
**Pull request by benchristel**
We've included an example implementation of the hook functions so you can see the intended usage of the new GPDB interface. We'll remove `idle_session_timeout_hook_example` before this PR is merged.

Co-authored-by: David Sharp <dsharp@pivotal.io>
Co-authored-by: Ben Christel <bchristel@pivotal.io>
Co-authored-by: Amil Khanzada <akhanzada@pivotal.io>

### CODE REVIEW

**Comment by amilkh** [comment]
Note this won't be merged to master (GPDB 6).
PG 9.3 introduced a timeout.c that may make this easier, so when it is time to add this hook to master, we will likely use timeout.c (WIP cherry-pick of timeout.c to 5X_STABLE: https://github.com/gp-workload-manager/gpdb/tree/cherry-pick-timeout-c)

**Comment by hlinnaka** [comment]
I don't understand how an extension author would use this hook. Some documentation would be nice. And/or a more realistic example implementation of the hook. What do you plan to use the hook for?

Why does the hook implementation need to duplicate the logic to check if gangs exist and kill them?

**Comment by amilkh** [comment]
Thanks for taking a look, Heikki!

`idle_session_timeout_hook` is intended to allow Postgres extension developers to define a function that will kill idle sessions after a certain timeout and/or various other conditions, e.g. current user or resource group. In our example implementation, all sessions will be killed after 5 seconds.

> Why does the hook implementation need to duplicate the logic to check if gangs exist and kill them?
This is because the SIGALARM will happen twice if `gp_vmem_idle_resource_timeout` is set to a value such as 2 seconds (reflected in `IdleSessionGangTimeout`). The first time the SIGALARM happens is after 2 seconds, `HandleClientWaitTimeout()` is called and the idle gangs are destroyed because `GangsExist()` is true. The second time the SIGALARM happens (after 5 - 2 = 3 seconds), `HandleClientWaitTimeout()` is called and since `GangsExist()` is false, the session will be killed.

**Comment by amilkh** [comment]
Another point is that the hook is not TDD'ed here because only <10 lines of code is going to merged to GPDB (the hook function declaration and invocations); the `idle_session_timeout_hook_example` will be in an extension that will be TDD'ed

**Comment by danielgustafsson** [comment]
Reviewing a hook PR is to a large extent about figuring out if the hook is in the right place, and without documentation for extension writers that's hard to do, so the documentation needs to be added first. The modes should also be given proper descriptive names, either as a `#define` or an enum. Taking the mode as an argument rather than having two functions seems fine on the face of it.

> Note this won't be merged to master (GPDB 6).

This gets a :-1: from me, this hook (and everything else) should be added to master first and then be backported to 5. It may well be that the implementation is different (and it may evolve over time), but the exposed API and functionality of the hook **should** remain for extension authors. If the hook will need to be changed as we merge 9.3 then so be it, we should never be in a position where an extension author can't target master. Adding a `GPDB_93_MERGE_FIXME` comment on the hook implementation ensures that we won't forget to update it when 9.3 is merged.

> Another point is that the hook is not TDD'ed here because only <10 lines of code is going to merged to GPDB

Unless impossible to orchestrate due to killing processes, a realistic test implementation should still be included, like how we have tests for other hooks. See `src/test/regress/query_info_hook_test/` and `src/test/regress/hooktest/` for inspiration.

**Comment by amilkh** [comment]
Thanks @danielgustafsson for the detailed feedback/insights!

> [...] The modes should also be given proper descriptive names, either as a #define or an enum. 
> [...] this hook (and everything else) should be added to master first and then be backported to 5.
> [...] Adding a GPDB_93_MERGE_FIXME comment

+1 on all of these!


> Unless impossible to orchestrate due to killing processes, a realistic test implementation should still be included, like how we have tests for other hooks. See src/test/regress/query_info_hook_test/ and src/test/regress/hooktest/ for inspiration.

Do we really need to do this? Is something like [resgroup_test.c](https://github.com/greenplum-db/gpdb/blob/master/src/backend/utils/resgroup/test/resgroup_test.c) sufficient?



> Reviewing a hook PR is to a large extent about figuring out if the hook is in the right place, and without documentation for extension writers that's hard to do, so the documentation needs to be added first.

What kind of documentation are you looking for? Perhaps the test coverage + code comments near the hooks could be good enough documentation?

**Comment by danielgustafsson** [comment]
> On 15 Mar 2018, at 17:18, Amil Khanzada <notifications@github.com> wrote:

> Do we really need to do this? Is something like resgroup_test.c sufficient?

Well, we need an adequate test to know that we aren’t breaking the intended usecase of the hook. If nothing is testing the hook on a regular basis, then we won’t know about breakage until we ship. I will leave defining “adequate” to you, but I don’t see how testing a hook in a mocked unit test would be done (well).

> What kind of documentation are you looking for? Perhaps the test coverage + code comments near the hooks could be good enough documentation?

Good question. We don’t really have a good place to add developer documentation like this, ideally we should start a section in the handbook for how to extend Greenplum, but thats a larger discussion. For now I think adding a section on how to use the hook (not a novel, just a description on why and how and when the hook can/should be used) to src/backend/cdb/dispatcher/README.md makes the most sense. If you have a suggestion for a more suitable location for a README, feel free to add it there instead.

**Comment by dsharp-pivotal** [comment]
@danielgustafsson Thanks for the comments.

We have actually made quite a big refactor, to move the idle session gang kill logic and our new idle session killing logic into separate files so we can do CMockery tests.

We think this makes the code easier to understand and test.

What do you think?

File: src/backend/tcop/idle_resource_cleaner.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+	/*
+	 * cancel the timer, as there is no reason we need it to go off again.
+	 */
+	disable_sig_alarm(false);
+
+	//TODO: determine which action to take (is this the first or second sigalarm trigger?)
+	idle_session_timeout_action_hook();
+	idle_gang_timeout_action();
+}
```
Comment (xiong-gang) [inline src/backend/tcop/idle_resource_cleaner.c]: Need a little bit clarification here.
By default, 18 seconds timeout will trigger `idle_gang_timeout_action`, if we add a new hook, for example, 5 seconds timeout and trigger the new hook function, are we expecting `idle_gang_timeout_action` go off 13 seconds later?

Comment (xiong-gang) [reply src/backend/tcop/idle_resource_cleaner.c]: If there are only one 'valid' timeout and its hook function. Would it be better to maintain 2 global variables by updating them when an extension is loaded?

**Comment by dsharp-pivotal** [comment]
On Mon, Mar 19, 2018 at 12:13 AM, xiong-gang <notifications@github.com>
wrote:

> *@xiong-gang* commented on this pull request.
> ------------------------------
>
> In src/backend/tcop/idle_resource_cleaner.c
> <https://github.com/greenplum-db/gpdb/pull/4690#discussion_r175345518>:
>
> > + *
> + */
> +void
> +HandleClientWaitTimeout(void)
> +{
> +	elog(DEBUG2,"HandleClientWaitTimeout");
> +
> +	/*
> +	 * cancel the timer, as there is no reason we need it to go off again.
> +	 */
> +	disable_sig_alarm(false);
> +
> +	//TODO: determine which action to take (is this the first or second sigalarm trigger?)
> +	idle_session_timeout_action_hook();
> +	idle_gang_timeout_action();
> +}
>
> Need a little bit clarification here.
> By default, 18 seconds timeout will trigger idle_gang_timeout_action, if
> we add a new hook, for example, 5 seconds timeout and trigger the new hook
> function, are we expecting idle_gang_timeout_action go off 13 seconds
> later?
>
This is a good point. We have mostly been thinking about our own use case,
where an idle session will kill the process. Because of this, we don't care
about the case when the session timeout is less than the gang timeout: the
process will die which "cleans up" the idle gangs anyway. With the code
as-is, idle_session_timeout_action_hook() would need to re-arm SIGALRM on
its first call, and kill the process on the second.

If we expand the possible use cases for the hook (say, monitoring and
logging idle sessions, which *won't* kill the session), we'll need to think
more about how to call them correctly. In that case, if the session timeout
is less than the gang timeout, then, yes, the gang timeout should be
re-armed here for 13 seconds.

File: src/backend/tcop/idle_resource_cleaner.c (hunk 1 above)
Comment (amilkh) [reply src/backend/tcop/idle_resource_cleaner.c]: Do you mean that if there are only two actions (session timeout and gang timeout), that it is better to use global variables rather than hooks to implement this logic?

We refactored out the gang and session killing timeout logic into two functions in a new file because it is easier to see it's doing two different things, and also this helps refactor GPDB to be easier to test.

**Comment by dsharp-pivotal** [comment]
@danielgustafsson we put idle_resource_cleaner.c into `src/backend/tcop` because postgres.c is there and uses it. However, proc.c also uses it, so is `src/backend/storage/lmgr` better?

File: src/backend/tcop/idle_resource_cleaner.c (hunk 1 above)
Comment (xiong-gang) [reply src/backend/tcop/idle_resource_cleaner.c]: I understand how it works now. Please ignore my comments above.
We should be careful in the `idle_session_timeout_action_hook`, for example, how do we determine it's an 'idle' session? If determine it with `GangsExists() == false`, then we should make sure `idle_gang_timeout_action` is triggered first, which means `idle_gang_timeout < idle_session_timeout_hook()`, but user can change idle_gang_timeout in a transaction block.

Comment (dsharp-pivotal) [reply src/backend/tcop/idle_resource_cleaner.c]: It doesn't matter to this code that the user can change idle_gang_timeout. When they do so (with a SET command), the session is no longer idle, DoingCommandRead becomes false (so HandleClientWaitTimeout doesn't get called), and new, non-idle SIGALRM timeouts get set. When the session goes idle again after running the SET command, we'll use the new timeout values.

File: src/include/tcop/idle_resource_cleaner.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#ifndef GPDB_IDLE_RESOURCE_CLEANER_H
+#define GPDB_IDLE_RESOURCE_CLEANER_H
```
Comment (danielgustafsson) [inline src/include/tcop/idle_resource_cleaner.h]: Please match header guard with actual filename, ie `IDLE_RESOURCE_CLEANER_H` in this case.

File: src/include/tcop/idle_resource_cleaner.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#ifndef GPDB_IDLE_RESOURCE_CLEANER_H
```
Comment (danielgustafsson) [inline src/include/tcop/idle_resource_cleaner.h]: Header with copyright stanza missing.

File: src/include/storage/proc.h [hunk 4]
```diff
@@ -245,7 +245,6 @@ extern int IdleSessionGangTimeout;
 
 extern volatile bool cancel_from_timeout;
 
-
```
Comment (danielgustafsson) [inline src/include/storage/proc.h]: Please avoid unrelated whitespace changes, however worthy of doing they are.

File: src/backend/tcop/test/idle_resource_cleaner_test.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+main(int argc, char *argv[])
+{
+	cmockery_parse_arguments(argc, argv);
+
+	const UnitTest tests[] = {
+		test_with_setup_and_teardown(test__MinPositiveInt),
+	};
+
+	run_tests(tests);
+}
```
Comment (danielgustafsson) [inline src/backend/tcop/test/idle_resource_cleaner_test.c]: This test seem pretty meaningless for the functionality discussed?

File: src/backend/tcop/idle_resource_cleaner.c (hunk 1 above)
Comment (amilkh) [reply src/backend/tcop/idle_resource_cleaner.c]: Also note that the idle session timeout can change at any time. We will think on how to solve that problem, but it's probably not necessary for this PR.

**Comment by Eulerizeit** [comment]
Was this merged into master already? Am I imagining that?

File: src/backend/tcop/idle_resource_cleaner.c [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		 * because it can't be running when we are idle.
+		 *
+		 * GPDB_93_MERGE_FIXME: replace the enable_sig_alarm() calls with the
+		 * new functionality provided in timeout.c from PG 9.3.
+		 *
+		 * We want the time value to be long enough so we don't free gangs prematurely.
+		 * This means giving the end user enough time to type in the next SQL statement
+		 *
+		 */
+		int sigalarm_timeout = MinPositiveInt(get_idle_gang_timeout(), get_idle_session_timeout_hook());
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: Since hooks can be unset, you should test it before dereferencing it.

```c
int sigalarm_timeout;
int session_timeout = -1;

if (get_idle_session_timeout_hook)
        session_timeout = (*get_idle_session_timeout_hook)();

sigalarm_timeout = MinPositiveInt(get_idle_gang_timeout(), session_timeout);
```
Also note that the PostgreSQL coding convention calls for explicit dereferencing of the function pointer, so we should do that too, as above, to keep code consistent.

Also see above comment about hook execution.

File: src/backend/tcop/idle_resource_cleaner.c [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	elog(DEBUG2,"HandleClientWaitTimeout");
+
+	/*
+	 * cancel the timer, as there is no reason we need it to go off again.
+	 */
+	disable_sig_alarm(false);
+
+	//TODO: determine which action to take (is this the first or second sigalarm trigger?)
+	idle_session_timeout_action_hook();
+	idle_gang_timeout_action();
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: See above, these hooks can be unset so needs to be explicitly tested before dereferencing.

File: src/backend/tcop/idle_resource_cleaner.c [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+void (*idle_session_timeout_action_hook)(void) = idle_session_timeout_action;
+
+
+static int MinPositiveInt(int x, int y)
+{
+	if (y <= 0) return x;
+	if (x <= 0) return y;
+
+	return Min(x, y);
+}
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: What about `MinPositiveInt(-1, -1);`? Since it's **PositiveInt**, sounds like it should `Assert()` that at least one parameter is non-negative.

Couldn't this be a macro, or at least inlined?

File: src/backend/tcop/idle_resource_cleaner.c [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	return 0;
+}
+
+static void idle_session_timeout_action()
+{
+	return;
+}
+
+int (*get_idle_session_timeout_hook)(void) = get_idle_session_timeout;
+void (*idle_session_timeout_action_hook)(void) = idle_session_timeout_action;
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: This is using hooks in a drastically different way from the rest of the codebase, and I'm not sure it's wise to mix concepts for extension authors. Here the hook is the expected codepath, and the core functionality is implemented using the hook. Normally what would be done is a conditional hook execution:
```c
if (get_idle_session_timeout_hook)
        session_timeout = (*get_idle_session_timeout_hook) ();
else
        session_timeout = get_idle_session_timeout();
```
The extension implementing the hook will then save the current value of the hook in a `previous_hook` variable, and execute that when leaving or call the original function if no other hook was defined. This way hooks can be daisychained such that they extend the core functionality without having to duplicate it. See below for an excerpt from `src/test/regress/hook_test/hook_test.c`:
```c
static planner_hook_type	prev_planner_hook = NULL;
...
void
_PG_init(void)
{
	prev_planner_hook = planner_hook;
	planner_hook = test_hook;
}
...
static PlannedStmt *
test_hook(Query *parse, int cursorOptions, ParamListInfo boundParams)
{
	PlannedStmt *stmt;
        ...
	if (prev_planner_hook)
		stmt = (* prev_planner_hook)(parse, cursorOptions, boundParams);
	else
		stmt = standard_planner(parse, cursorOptions, boundParams);
	return stmt;
}
```
For this hook, daisychaining may not make sense but conditionally executing and leaving that decision to the hook author seems a better bet. Also, dereferencing a hook that might not be set is usually not a good idea.

File: src/backend/tcop/idle_resource_cleaner.c [hunk 10]
```diff
@@ -0,0 +1,134 @@
+#include <postgres.h>
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: Header with copyright stanza missing.

File: src/backend/storage/lmgr/proc.c [hunk 11]
```diff
@@ -1668,68 +1670,6 @@ disable_sig_alarm(bool is_statement_timeout)
...
-			 * we don't need to do anything special.  Specifically, we DON'T want
-			 * to act like we are now in a new session, since that would be confusing
-			 * in the log.
-			 *
-			 */
-			DisconnectAndDestroyAllGangs(false);
-		}
-	}
-}
-
```
Comment (danielgustafsson) [inline src/backend/storage/lmgr/proc.c]: When removing this you can remove the inclusion of `cdbgang.h` from `proc.c` too right?

File: src/backend/storage/lmgr/proc.c [hunk 12]
```diff
@@ -64,6 +65,7 @@
 #include "cdb/cdbvars.h"  /*Gp_is_writer*/
 #include "port/atomics.h"
 #include "utils/session_state.h"
+#include "tcop/idle_resource_cleaner.h"
 
 /* GUC variables */
```
Comment (danielgustafsson) [inline src/backend/storage/lmgr/proc.c]: Can we move `IdleSessionGangTimeout` to the new file too?

File: src/backend/tcop/idle_resource_cleaner.c [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+#include <cdb/cdbvars.h>
+#include <cdb/cdbgang.h>
+#include <storage/proc.h>
+#include <catalog/namespace.h>
+#include <tcop/idle_resource_cleaner.h>
+
+static int get_idle_gang_timeout(void)
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: The coding convention is:
```c
static int
get_idle_gang_timeout(void)
...
```

File: src/backend/tcop/postgres.c [hunk 14]
```diff
@@ -4883,29 +4884,7 @@ PostgresMain(int argc, char *argv[],
...
-			 *
-			 * We want the time value to be long enough so we don't free gangs prematurely.
-			 * This means giving the end user enough time to type in the next SQL statement
-			 *
-			 */
-			if (IdleSessionGangTimeout > 0 && GangsExist())
-				if (!enable_sig_alarm( IdleSessionGangTimeout /* ms */, false))
-					elog(FATAL, "could not set timer for client wait timeout");
-		}
+		CleanupIdleResources();
```
Comment (danielgustafsson) [inline src/backend/tcop/postgres.c]: This is a fairly hot codepath, I wonder if it's worth keeping the check for `GP_ROLE_DISPATCH` to avoid a jump on the segments?

**Comment by danielgustafsson** [comment]
> On 20 Mar 2018, at 01:36, David Sharp <notifications@github.com> wrote:
> 
> @danielgustafsson we put idle_resource_cleaner.c into src/backend/tcop because postgres.c is there and uses it. However, proc.c also uses it, so is src/backend/storage/lmgr better?

I think tcop is the right place.

**Comment by danielgustafsson** [comment]
> On 16 Mar 2018, at 01:36, David Sharp <notifications@github.com> wrote:
> 
> @danielgustafsson Thanks for the comments.
> 
> We have actually made quite a big refactor, to move the idle session gang kill logic and our new idle session killing logic into separate files so we can do CMockery tests.
> 
> We think this makes the code easier to understand and test.
> 
> What do you think?

I like the refactoring, but I think it will be hard to mock a test for this. I’m happy to be proven wrong, but I would’ve gone for a simpler extension test like the ones we already have in ICW.

**Comment by danielgustafsson** [comment]
Another note, this PR is targeting `5X_STABLE` which is a big **No**, it needs to be re-pointed against `master` and only backported from there once merged.

File: src/backend/tcop/idle_resource_cleaner.c (hunk 8 above)
Comment (dsharp-pivotal) [reply src/backend/tcop/idle_resource_cleaner.c]: See the test. If there are no positive integers, then it should return a non-positive integer. This matches what we want: if there are no enabled (positive) timeouts, then don't enable SIGALRM. If you have a name that makes this clearer, we can use it. Otherwise, I think this deserves a comment, which we can add.

As for macro: no thanks, for the usual reasons. And for inlining: the compiler can decide (`inline` doesn't force the compiler to inline anyway).

**Comment by dsharp-pivotal** [comment]
>Another note, this PR is targeting `5X_STABLE` which is a big **No**, it needs to be re-pointed against `master` and only backported from there once merged.

ACK. We're developing it on 5X_STABLE, but we'll forward port it and push it to master first.

It's kind of an odd situation: We won't actually need this on master, because by the time 6.0 is released, you'll have merged timeout.c from Postgres 9.3, which makes a good amount of this unnecessary. (We'll still need a hook in CleanupIdleResources() to let us set our timeout, but that's it. The other "action" hook will just be the custom timeout's handler).

**Comment by danielgustafsson** [comment]
> On 20 Mar 2018, at 22:24, David Sharp <notifications@github.com> wrote:

> See the test. If there are no positive integers, then it should return a non-positive integer. This matches what we want: if there are no enabled (positive) timeouts, then don't enable SIGALRM. If you have a name that makes this clearer, we can use it. Otherwise, I think this deserves a comment, which we can add.

I would like another name, but I don’t have any good suggestions right now. MinTimeout()? A comment should be added regardless though, to ensure it’s clear.

File: src/include/tcop/idle_resource_cleaner.h (hunk 2 above)
Comment (benchristel) [reply src/include/tcop/idle_resource_cleaner.h]: Done

File: src/include/tcop/idle_resource_cleaner.h [hunk 15]
```diff
@@ -0,0 +1,28 @@
+/*-------------------------------------------------------------------------
+ *
+ * idle_resource_cleaner.h
+ *	  Functions used to clean up resources on idle sessions.
+ *
+ *
+ * Portions Copyright (c) 2012-Present Pivotal Software, Inc.
```
Comment (benchristel) [inline src/include/tcop/idle_resource_cleaner.h]: @danielgustafsson is this the right copyright header, or should we include a line about `Greenplum, Inc.` similar to `resgroup.h`?
Also should it be `2018-Present`?

File: src/include/storage/proc.h (hunk 4 above)
Comment (benchristel) [reply src/include/storage/proc.h]: Fixed

File: src/backend/tcop/idle_resource_cleaner.c (hunk 9 above)
Comment (amilkh) [reply src/backend/tcop/idle_resource_cleaner.c]: Thanks, we dereference and now default to NULL.

However, we won't do daisychaining here since it doesn't make much sense (but it can be an added feature in future if people want it on this hook).

File: src/backend/tcop/postgres.c [hunk 16]
```diff
@@ -5132,37 +5133,16 @@ PostgresMain(int argc, char *argv[],
 
 		/*
 		 * (2b) Check for temp table delete reset session work.
+		 * Also clean up idle resources.
 		 */
-		if (Gp_role == GP_ROLE_DISPATCH)
+		if (Gp_role == GP_ROLE_DISPATCH) {
 			CheckForResetSession();
+			StartIdleResourceCleanupTimers();
+		}
```
Comment (amilkh) [inline src/backend/tcop/postgres.c]: > This is a fairly hot codepath, I wonder if it's worth keeping the check for GP_ROLE_DISPATCH to avoid a jump on the segments?
@danielgustafsson we've moved the function check into the if block.

Also we combined it with the other gpdb-specific (2b) section code.

Note that upstream postgres only has a (2) and (3) section.
https://github.com/postgres/postgres/blob/master/src/backend/tcop/postgres.c
https://github.com/greenplum-db/gpdb/blob/master/src/backend/tcop/postgres.c

**Comment by amilkh** [comment]
@danielgustafsson thanks for the detailed review and comments!

We've addressed most of them.

Our next steps are to:
1. Keep TDD'ing the code with CMockery unit tests. We may possibly change our logic to be simpler regarding the timeout checks / which sigalarm to set first.
2. Consider adding an icw hook test (like hook_test.c)
3. Rebase and squash into a giant commit

File: src/include/tcop/idle_resource_cleaner.h (hunk 15 above)
Comment (danielgustafsson) [reply src/include/tcop/idle_resource_cleaner.h]: Since this is copying existing code which AFAICT was present in the Greenplum Inc days, there should be a `Portions Copyright ..` stanza here as well. For the Pivotal line it should cover from when Pivotal took over the IP for the code, as this is existing code.

File: src/backend/tcop/idle_resource_cleaner.c [hunk 17]
```diff
@@ -0,0 +1,134 @@
+/*-------------------------------------------------------------------------
+ *
+ * idle_resource_cleaner.c
+ *
+ *
+ * Portions Copyright (c) 2012-Present Pivotal Software, Inc.
+ *
+ *
+ * IDENTIFICATION
+ *	    src/include/tcop/idle_resource_cleaner.c
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: Copy-pasteo, this should be `src/backend/tcop/idle_resource_cleaner.c`

File: src/backend/tcop/Makefile [hunk 18]
```diff
@@ -12,7 +12,8 @@ subdir = src/backend/tcop
 top_builddir = ../../..
 include $(top_builddir)/src/Makefile.global
 
-OBJS= dest.o fastpath.o postgres.o pquery.o utility.o
+OBJS= dest.o fastpath.o postgres.o pquery.o utility.o \
+	idle_resource_cleaner.o
```
Comment (danielgustafsson) [inline src/backend/tcop/Makefile]: Thinking more about this, can we split this into two blocks to ease the merge effort? Something like this will ensure we avoid conflicts around the OBJS line:
```make
OBJS= dest.o fastpath.o postgres.o pquery.o utility.o

# Greenplum objects
OBJS+= idle_resource_cleaner.o
```

File: src/backend/tcop/idle_resource_cleaner.c (hunk 8 above)
Comment (danielgustafsson) [reply src/backend/tcop/idle_resource_cleaner.c]: :+1: on the rename to `MinTimeout()` and the comment,  thanks!

File: src/backend/tcop/idle_resource_cleaner.c [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+	else
+		IdleSessionTimeout = 0;
+
+	int sigalarm_timeout = MinTimeout(get_idle_gang_timeout(),
+	                                  IdleSessionTimeout);
+
+	if (sigalarm_timeout > 0)
+	{
+		if (!enable_sig_alarm(sigalarm_timeout, false))
+			elog(FATAL, "could not set timer for client wait timeout");
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: While not introduced by this patch, I wonder why we consider this a `FATAL`? Seems a big gun to use, but I might be missing something. Not to be changed in this PR however, just thinking out loud.

**Comment by danielgustafsson** [comment]
> On 21 Mar 2018, at 00:39, Amil Khanzada <notifications@github.com> wrote:
> 
> @danielgustafsson thanks for the detailed review and comments!
> 
> We've addressed most of them.

Thanks, it’s looking much better already, good work!

**Comment by danielgustafsson** [comment]
> On 20 Mar 2018, at 22:34, David Sharp <notifications@github.com> wrote:
> 
> Another note, this PR is targeting 5X_STABLE which is a big No, it needs to be re-pointed against master and only backported from there once merged.
> 
> ACK. We're developing it on 5X_STABLE, but we'll forward port it and push it to master first.
> 
Cool, just wanted to double-check.
> It's kind of an odd situation: We won't actually need this on master, because by the time 6.0 is released, you'll have merged timeout.c from Postgres 9.3, which makes a good amount of this unnecessary. (We'll still need a hook in CleanupIdleResources() to let us set our timeout, but that's it. The other "action" hook will just be the custom timeout's handler).
> 
I know, this is an odd situation where we will most likely replace all of this before shipping 6.0. Thats fine though, that’s a cost incurred by working in a downstream fork which is actively merged. Adding a GPDB_XY_MERGE_FIXME comment (like you’ve done) takes care of it.

File: src/include/tcop/idle_resource_cleaner.h (hunk 15 above)
Comment (benchristel) [reply src/include/tcop/idle_resource_cleaner.h]: @amilkh is saying that maybe we don't need to put The `Greenplum Inc.` header because although the `HandleClientWaitTimeout()` was pulled out of proc.c and is old code associated with the gp_vmem idle gang killer GUC, everything has been re-gutted and is almost like new code.

File: src/backend/tcop/idle_resource_cleaner.c (hunk 19 above)
Comment (amilkh) [reply src/backend/tcop/idle_resource_cleaner.c]: Hmm, well looking into the code of `enable_sigalarm` we see it can only return false if setitimer fails.

```c
	/* If we reach here, okay to set the timer interrupt */
	MemSet(&timeval, 0, sizeof(struct itimerval));
	timeval.it_value.tv_sec = delayms / 1000;
	timeval.it_value.tv_usec = (delayms % 1000) * 1000;
	if (setitimer(ITIMER_REAL, &timeval, NULL))
		return false;
```


setitimer only returns non-zero in these cases (from https://www.systutorials.com/docs/linux/man/2-setitimer/):
```
*RETURN VALUE
On success, zero is returned. On error, -1 is returned, and errno is set appropriately.  

*ERRORS
EFAULT
new_value, old_value, or curr_value is not valid a pointer.
EINVAL
which is not one of ITIMER_REAL, ITIMER_VIRTUAL, or ITIMER_PROF; or (since Linux 2.6.22) one of the tv_usec fields in the structure pointed to by new_value contains a value outside the range 0 to 999999
```

Looks like this is very unlikely to happen, unless something like a negative number is passed in as the timeout or Bad Stuff™ happens in the system, so FATAL is probably OK since we want to fail hard so the bad backend dies (as it doesn't do the proper behavior).

**Comment by danielgustafsson** [comment]
> On 21 Mar 2018, at 18:35, Ben Christel <notifications@github.com> wrote:

> @amilkh is saying that maybe we don't need to put The Greenplum Inc. header because although the HandleClientWaitTimeout() was pulled out of proc.c and is old code associated with the gp_vmem idle gang killer GUC, everything has been re-gutted and is almost like new code.

Doesn’t matter, their Copyright still applies and the mention must be retained.

File: src/backend/tcop/idle_resource_cleaner.c [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+ * We want the time value to be long enough so we don't free gangs prematurely.
+ * This means giving the end user enough time to type in the next SQL statement
+ *
+ *
+ * GPDB_93_MERGE_FIXME: replace the enable_sig_alarm() calls with the
+ * new functionality provided in timeout.c from PG 9.3.
+ */
+void StartIdleResourceCleanupTimers()
+{
+//	CreateSchedule(0, 0, &CurrentIdleResourceCleanupSchedule);
```
Comment (paul-guo-) [inline src/backend/tcop/idle_resource_cleaner.c]: I saw two comment styles // and /* */ mixed in the patch. I'd suggest using /* */ for C code in the whole patch although // is accepted by gcc.

Comment (benchristel) [reply src/backend/tcop/idle_resource_cleaner.c]: Thanks @paul-guo-! We will fix.

File: src/backend/tcop/idle_resource_cleaner.c (hunk 19 above)
Comment (danielgustafsson) [reply src/backend/tcop/idle_resource_cleaner.c]: Right, we should at least change the error message to include errno and `%m`, but not in this PR.

File: src/test/regress/hooktest/hook_test.c [hunk 21]
```diff
@@ -3,26 +3,11 @@
...
-{
-	prev_planner_hook = planner_hook;
-	planner_hook = test_hook;
-}
+PG_MODULE_MAGIC;
 
-void
-_PG_fini(void)
-{
-	planner_hook = prev_planner_hook;
```
Comment (danielgustafsson) [inline src/test/regress/hooktest/hook_test.c]: This is still required in `_PG_fini()`, even if we don't use multiple modules here it's still worth doing the right thing to avoid subtle bugs if we at some point do.

**Comment by amilkh** [comment]
We have made two commits on the PR right now:
1. refactor idle gang killing logic to a new file with Cmockery tests
2. add a (failing) WIP icw test for the new idle session timeout/killing hook

danielgustafsson can you please give us your thoughts on no. 1 and no.2?
Specifically, for the icw test:
a. should we use a new file (not hooktest.c)?
b. it looks like there are two existing hook tests already (GNUmakefile says `installcheck-good: all twophase_pqexecparams hooktest query_info_hook_test`). Why are they separate, i.e. should query_info_hook_test go into hooktest.c?
c. any ideas of what else we should test about the idle session timeout hook in icw? We were thinking this is enough since we will have Cmockery unit test coverage and don't want to do timeouts/sleeps in icw.

-Amil and David

**Comment by benchristel** [comment]
@danielgustafsson 
We believe the idle session timeout/killing hook infrastructure is mostly complete!

We will want to refactor the ICW test so all 3 hook tests are in one file or each test has its own file, but the functionality for testing our feature is added and all our tests pass locally (but fail on PR pipeline).

Could you please let us know your final comments/review and perhaps a "conditional" +1?

-Ben and Amil

File: src/backend/tcop/idle_resource_cleaner.c [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+ * Note we don't need to worry about the statement timeout timer
+ * because it can't be running when we are idle.
+ *
+ * We want the time value to be long enough so we don't free gangs prematurely.
+ * This means giving the end user enough time to type in the next SQL statement
+ *
+ * GPDB_93_MERGE_FIXME: replace the enable_sig_alarm() calls with the
+ * new functionality provided in timeout.c from PG 9.3.
+ */
+void StartIdleResourceCleanupTimers()
```
Comment (danielgustafsson) [inline src/backend/tcop/idle_resource_cleaner.c]: Mixing C style in the same file, `void` should go its own line.

File: src/include/tcop/idle_resource_cleaner.h (hunk 15 above)
Comment (danielgustafsson) [reply src/include/tcop/idle_resource_cleaner.h]: As commented elsewhere, that doesn't matter and the copyright statement must be there.

**Comment by amilkh** [comment]
We've added the icw hook test and squashed all the idle session timeout logic into one commit and think this feature is complete!

Please let us know any final comments before we merge this in and thanks for reviewing!

@xiong-gang @paul-guo-
